src/axl.c

```diff
@@ -1376,12 +1410,13 @@ int AXL_Free (int id)
 
 int AXL_Stop ()
 {
+    const unsigned int kvtrees_count = axl_kvtrees_count();
```

Comment:
I believe you'll need to hold the lock for the the entirety of this function to prevent `axl_kvtree_count()` from changing

Reply:
Good point. For `AXL_Stop` the issue may be bit more complex.  I think if `count` changes (keeping in mind that it can only increase) during this function I am in deep trouble to begin with since there would be active transfers left after `AXL_Stop`. On the plus side tight now `count` is only modified by the main thread, so there is not chance of it changing. If I wanted to make `AXL_Stop` thread safe, I would need to record a "stopped" flag to prevent new transfers from being created once `AXL_Stop` starts executing.

Reply:
Looking at this I may be able to avoid the lock `axl_kvtrees` completely. The only reason it is there is that he pthreads only record the transfer ID integer and thus have to access the array to get the `file_list` pointer. However if one instead stored the `file_list` pointer (which never changes) in the `pdata` structure (in addition to the transfer ID) then one does not have multiple threads accessing `axl_kvtrees` and no lock is needed.

Reply:
I made the change to copy `file_list` in ad3e3c4c59fcbc09f3cc4061e741b9f77b877f46 which avoids this issue altogether.